File: 프로그래머스/LV2/코딩테스트 연습/연습문제/숫자 카드 나누기/숫자 카드 나누기.py
from math import gcd
from functools import reduce

def func(arr1, arr2):
    g = reduce(gcd, arr1)
    for v in arr2:
        if v % g == 0:
            break
    else:
        return g
    return 0
                
def solution(arrayA, arrayB):
    return max(func(arrayA, arrayB), func(arrayB, arrayA))

File: 프로그래머스/LV2/코딩테스트 연습/연습문제/숫자 카드 나누기/숫자 카드 나누기.cpp
#include <vector>
using namespace std;

int gcd(int a, int b)
{
    return b ? gcd(b, a % b) : a;
}

int GcdOfArray(vector<int>& arr, int idx)
{
    if (idx == arr.size() - 1) {
        return arr[idx];
    }
    int a = arr[idx];
    int b = GcdOfArray(arr, idx + 1);
    return gcd(a, b);
}

int func(vector<int> arr1, vector<int> arr2)
{
    bool f = true;
    int g = GcdOfArray(arr1, 0);
    for(int v : arr2)
    {
        if(v % g == 0)
        {
            f = false;
            break;
        }
    }
    return f ? g : 0;
}

int solution(vector<int> arrayA, vector<int> arrayB)
{
    return max(func(arrayA, arrayB), func(arrayB, arrayA));
}